File: hospital_bed_management/hospital_bed_management/page/patient_allotment_an/patient_allotment_an.js
frappe.pages['patient-allotment-an'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Patient Allotment and Updation',
		single_column: true
	});

	$("<div class='allot-patients' style='min-height: 200px; padding:15px; height:400px; overflow:auto;'>\
		</div>").appendTo(page.main);
	wrapper.patient_allotment_an = new frappe.PatientAllotment(wrapper);
}

frappe.PatientAllotment = Class.extend({
	init: function(wrapper) {
		this.wrapper = wrapper;
		this.body = $(this.wrapper).find(".allot-patients");
		this.filters = {};
		this.add_filters();
		// this.search_hospital();
		this.allot_bed_to_patient();
	},

	add_filters: function(){
		var me = this;

		frappe.call({
			method:"hospital_bed_management.hospital_bed_management.page.patient_allotment_an.patient_allotment_an.get_all_patients",
			callback: function(r) {
				if(r.message){
					me.options = r.message;
					me.filters.patient_name = me.wrapper.page.add_field({
						fieldname: "patient_name",
						label: __("Patient Name"),
						fieldtype: "Data"
					});
					$(me.wrapper).find('[data-fieldname = patient_name]').autocomplete({source: r.message})
				}

				me.filters.patient_type = me.wrapper.page.add_field({
					fieldname: "patient_type",
					label: __("Type of Patients"),
					fieldtype: "Select",
					options: (["","Indigent","Weaker"])
				});

				me.filters.allotment_id = me.wrapper.page.add_field({
					fieldname: "allotment_id",
					label: __("Provisional Allotment Id"),
					fieldtype: "Link",
					options: ("Patient Allotment")
				});

				me.filters.status = me.wrapper.page.add_field({
					fieldname: "status",
					label: __("Status"),
					fieldtype: "Select",
					options: (["Recommended","Rejected","Alloted"])
				});

				me.search = me.wrapper.page.add_field({
					fieldname: "search",
					label: __("Search"),
					fieldtype: "Button"
				});

				$('select option:contains("Recommended")').prop('selected',true);

				me.search_hospital();
			}
		});
	},

	search_hospital: function(){
		var me = this
		me.search.$input.on("click", function() {
			$('.table').remove();
			me.recommended_hospital_details();
		});
	},

	recommended_hospital_details: function(){
		var me = this;
		patient_name = this.filters.patient_name.$input.val();
		p_type =  this.filters.patient_type.$input.val();
		allotment_id =  this.filters.allotment_id.$input.val();
		status =  this.filters.status.$input.val();

		frappe.call({
			method:"hospital_bed_management.hospital_bed_management.page.patient_allotment_an.patient_allotment_an.get_recommendation_detials",
			args:{
				"patient_name": patient_name,
				"p_type": p_type,
				"allotment_id": allotment_id,
				"status": status,
			},
			callback: function(r) {
				if(r.message){
					$(me.wrapper).find(".allot-patients").html(frappe.render_template("patient_allotment_an", {"data":r.message}))
				}

				me.allot_bed_to_patient();
				me.reject_allotment();
			}
		});
	},

	allot_bed_to_patient: function(){
		var me = this
		$(this.wrapper).find('.btn-allot').on("click", function() {
			if($("input:radio[name='select']").is(":checked")!=true){
				frappe.throw(__("Please select Patient first to Bed Allotment..."));
			}
			else{ 
				if(me.filters.status.$input.val()=="Recommended"){
					var patient_allot_id = $("input[name=select]:checked").closest('tr').attr('allot-id')
					frappe.call({
						method:"hospital_bed_management.hospital_bed_management.page.patient_allotment_an.patient_allotment_an.update_hospital_beds_availability",
						args:{
							"allotment_id":patient_allot_id
						},
						callback: function(r) {
							msgprint("Bed Alloted Successfully to this patient...")
							window.location.reload();
						}
					});
				}
				else if(me.filters.status.$input.val()=="Alloted"){
					frappe.throw(__("Already alloted Bed to selected patient."));
				}
				else{
					frappe.throw(__("Selected patient status must be Recommended for Bed Allotment."));
				}
			}
		})
	},

	reject_allotment: function(){
		var me = this
		$(this.wrapper).find('.btn-reject').on("click", function() {
			if($("input:radio[name='select']").is(":checked")!=true){
				frappe.throw(__("Please select Patient first for Allotment rejection..."));
			}
			else{ 
				if(me.filters.status.$input.val()=="Recommended"){
					var patient_allot_id = $("input[name=select]:checked").closest('tr').attr('allot-id')
					frappe.call({
						method:"hospital_bed_management.hospital_bed_management.page.patient_allotment_an.patient_allotment_an.reject_bed_allotment",
						args:{
							"allotment_id":patient_allot_id
						},
						callback: function(r) {
							msgprint("Bed Allocation Rejected to this patient...")
							window.location.reload();
						}
					});
				}
				else if(me.filters.status.$input.val()=="Alloted"){
					frappe.throw(__("Already alloted Bed to selected patient."));
				}
				else{
					frappe.throw(__("Bed Allotment is already rejected for selected patient."));
				}
			}
		})
	}
})

File: hospital_bed_management/hospital_bed_management/page/patient_allotment_an/patient_allotment_an.py
from __future__ import unicode_literals
import frappe
from frappe import throw, _

@frappe.whitelist()
def get_recommendation_detials(patient_name,p_type,allotment_id,status):
	conditions = ""
	if patient_name: conditions += "and patient_name = '%s'"%(patient_name)
	if p_type: conditions += "and patient_type = '%s' "%(p_type)
	if allotment_id: conditions += "and name = '%s'"%(allotment_id)
	if status: conditions += "and status = '%s'"%(status)

	val = frappe.db.sql("""select defvalue from `tabDefaultValue` where parent = '%s' and defkey = 'Hospital Registration' """%(frappe.session.user), as_dict=1)
	if val:
		data = frappe.db.sql("""select * from `tabPatient Allotment` where hospital_name = '%s' %s """%(val[0]['defvalue'],conditions),as_dict=1)
	else:
		frappe.throw(_("Please set Hospital in User Permissions first."))
	for d in data:
		files = ""
		documents = frappe.db.sql("""select document_name from `tabIncome Documents` where parent = '%s' """%(d["name"]),as_dict=1)
		for doc in documents:
			files = files + " " + doc['document_name'] + ","

		d['file_name'] = files[:-1]

	return data

@frappe.whitelist()
def get_all_patients():
	patients = frappe.db.sql("""select distinct patient_name from `tabPatient Allotment` """,as_list=1)
	p_list = []
	p_list = [p[0] for p in patients]
	return p_list

@frappe.whitelist()
def update_hospital_beds_availability(allotment_id):
	hospital = frappe.db.get_value("Patient Allotment", allotment_id, ["hospital_name","patient_type"],as_dict=True)
	
	if hospital['patient_type']=="Indigent" :
		i_alloted = frappe.db.get_value("Hospital Registration", hospital['hospital_name'], ["i_patient_alloted","i_available"],as_dict=True)
		if i_alloted['i_available'] != 0:
			allot = i_alloted['i_patient_alloted'] + 1
			total_i_beds = i_alloted['i_available'] - 1

			hosp = frappe.get_doc('Hospital Registration',hospital['hospital_name'])
			hosp.i_patient_alloted = allot
			hosp.i_available = total_i_beds
			hosp.flags.ignore_permissions = 1
			hosp.save()
		else:
			frappe.throw(_("Sorry...Beds are not available for Indigent patients."))
	elif hospital['patient_type']=="Weaker" :
		w_alloted = frappe.db.get_value("Hospital Registration", hospital['hospital_name'], ["w_patient_alloted","w_available"],as_dict=True)
		if w_alloted['w_available'] != 0:
			allot = w_alloted['w_patient_alloted'] + 1
			total_w_beds = w_alloted['w_available'] - 1

			hosp = frappe.get_doc('Hospital Registration',hospital['hospital_name'])
			hosp.w_patient_alloted = allot
			hosp.w_available = total_w_beds
			hosp.flags.ignore_permissions = 1
			hosp.save()
		else:
			frappe.throw(_("Sorry...Beds are not available for Weaker patients."))

	p_allotment_status = frappe.db.get_value("Patient Allotment", allotment_id, "Status")
	allotment = frappe.get_doc('Patient Allotment',allotment_id)
	allotment.status = "Alloted"
	allotment.save()

@frappe.whitelist()
def reject_bed_allotment(allotment_id):
	# hospital = frappe.db.get_value("Patient Allotment", allotment_id, ["hospital_name","patient_type"],as_dict=True)
	p_allotment_status = frappe.db.get_value("Patient Allotment", allotment_id, "Status")
	allotment = frappe.get_doc('Patient Allotment',allotment_id)
	allotment.status = "Rejected"
	allotment.save()